include/mfmg/amge.templates.hpp

```diff
@@ -0,0 +1,200 @@
+
+  // The std::move inhibits copy elision but the code does not work otherwise
+  return std::make_tuple(std::move(agglomerate_triangulation),
+                         agglomerate_to_global_tria_map);
+}
+
+template <int dim, typename VectorType>
+void AMGe<dim, VectorType>::output(std::string const &filename)
+{
+  dealii::DataOut<dim> data_out;
+  data_out.attach_dof_handler(_dof_handler);
+
+  unsigned int const n_active_cells =
+      _dof_handler.get_triangulation().n_active_cells();
+  dealii::Vector<float> subdomain(n_active_cells);
+  for (unsigned int i = 0; i < n_active_cells; ++i)
+    subdomain(i) = _dof_handler.get_triangulation().locally_owned_subdomain();
+  data_out.add_data_vector(subdomain, "subdomain");
+
+  dealii::Vector<float> agglomerates(n_active_cells);
+  unsigned int n = 0;
+  for (auto cell : _dof_handler.active_cell_iterators())
+  {
+    if (cell->is_locally_owned())
+      agglomerates(n) = cell->user_index();
+    ++n;
+  }
+  data_out.add_data_vector(agglomerates, "agglomerates");
+
+  data_out.build_patches();
+
+  std::string full_filename =
+      filename +
+      std::to_string(
+          _dof_handler.get_triangulation().locally_owned_subdomain());
+  std::ofstream output((full_filename + ".vtu").c_str());
+  data_out.write_vtu(output);
+
+  if (dealii::Utilities::MPI::this_mpi_process(_comm) == 0)
+  {
+    unsigned int const comm_size =
+        dealii::Utilities::MPI::n_mpi_processes(_comm);
+    std::vector<std::string> full_filenames;
+    for (unsigned int i = 0; i < comm_size; ++i)
+      full_filenames.push_back(filename + std::to_string(i) + ".vtu");
+    std::ofstream master_output(filename + ".pvtu");
+    data_out.write_pvtu_record(master_output, full_filenames);
+  }
+}
+
+template <int dim, typename VectorType>
+void AMGe<dim, VectorType>::setup(
+    std::array<unsigned int, dim> const &agglomerate_dim)
+{
+  // Flag the cells to build agglomerates.
+  unsigned int const n_agglomerates = build_agglomerates(agglomerate_dim);
+
+  // Parallel part of the setup.
+  std::vector<unsigned int> agglomerate_ids(n_agglomerates);
+  std::iota(agglomerate_ids.begin(), agglomerate_ids.end(), 1);
+  dealii::WorkStream::run(agglomerate_ids.begin(), agglomerate_ids.end(), *this,
+                          &AMGe::local_worker, &AMGe::copy_local_to_global,
+                          ScratchData(), CopyData());
+}
+
+template <int dim, typename VectorType>
+void AMGe<dim, VectorType>::local_worker(
+    std::vector<unsigned int>::iterator const &agg_id, ScratchData &,
+    CopyData &)
+{
+  dealii::Triangulation<dim> agglomerate_triangulation;
+  std::map<typename dealii::Triangulation<dim>::active_cell_iterator,
+           typename dealii::DoFHandler<dim>::active_cell_iterator>
+      agglomerate_to_global_tria_map;
+
+  std::tie(agglomerate_triangulation, agglomerate_to_global_tria_map) =
+      build_agglomerate_triangulation(*agg_id);
```

Comment:
Am I right that you build Triangulations and mappings but do not store them, so they simply go away?

Reply:
For now yes. I will use them when I do the local assembly to compute the eigenvectors.